src/validate.ts

```diff
@@ -26,18 +40,103 @@ function responseValidation(apiResponse: object, schema: object) {
 
 /**
  * Identifies the response type and validates the function based on the schemas.
- * @param apiResponse User defined api response
- * @param responseType User defined intent response
+ * @param req Request from one of the generate request functions.
+ * @param apiResponse User defined api response.
+ * @param syncData? Optional parameter to help identify what schemas to use for trait validation 
  * @return Errors from AJV validation, if any. Undefined otherwise.
  */
-export function validate(intentRequest: object, apiResponse: object) {
+export function validate(intentRequest: object, apiResponse: object, syncData?: object){
   const responseType = intentRequest['inputs'][0]['intent'];
+
   if (responseType === 'action.devices.SYNC') {
-    return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+    const validateSyncAPI = responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA); 
+    const syncErrors : object[] = [];
+    
+    if (validateSyncAPI) {
+        syncErrors.push(...validateSyncAPI);
+        return syncErrors;
+    }
+
+    const syncDevices = apiResponse['payload']['devices'];
+    const syncDevicesLength = syncDevices.length;
+    for (let i = 0; i < syncDevicesLength; i++) {
+      const traits = syncDevices[i]['traits'];
+      const attributes = syncDevices[i]['attributes'] || {};
+      for (let j = 0; j < traits.length; j++) {
+        const trait = traits[j];
+        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
+          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+          if (validateTraitRes) {
+            syncErrors.push(...validateTraitRes);
+          }
+        }
+      }
+   } 
+    return syncErrors.length ? syncErrors : undefined;
```

Comment:
I'd run a lint check as I think there may be a few style comments with the indentation here.